framework/src/util/Properties.cpp

```diff
@@ -24,99 +24,140 @@
 
 #include <limits>
 #include <stdexcept>
-#ifdef US_PLATFORM_WINDOWS
-#  include <string.h>
-#  define ci_compare strnicmp
-#else
-#  include <strings.h>
-#  define ci_compare strncasecmp
-#endif
+#include <utility>
+
+#include "PropsCheck.h"
 
 US_MSVC_PUSH_DISABLE_WARNING(4996)
 
 namespace cppmicroservices {
 
 const Any Properties::emptyAny;
 
-Properties::Properties(const AnyMap& p)
+void Properties::PopulateCaseInsensitiveLookupMap()
 {
-  if (p.size() > static_cast<std::size_t>(std::numeric_limits<int>::max())) {
-    throw std::runtime_error("Properties contain too many keys");
-  }
-
-  keys.reserve(p.size());
-  values.reserve(p.size());
-
-  for (auto& iter : p) {
-    if (Find_unlocked(iter.first) > -1) {
-      std::string msg("Properties contain case variants of the key: ");
-      msg += iter.first;
-      throw std::runtime_error(msg.c_str());
+  if (props.GetType() == AnyMap::ORDERED_MAP) {
+    for (auto itr = props.beginOM_TypeChecked();
+         itr != props.endOM_TypeChecked();
+         ++itr) {
+      caseInsensitiveLookup[props_check::ToLower(itr->first)] = itr->first;
+    }
+  } else if (props.GetType() == AnyMap::UNORDERED_MAP) {
+    for (auto itr = props.beginUO_TypeChecked();
+         itr != props.endUO_TypeChecked();
+         ++itr) {
+      caseInsensitiveLookup[props_check::ToLower(itr->first)] = itr->first;
     }
-    keys.push_back(iter.first);
-    values.push_back(iter.second);
+  } else {
+    throw std::runtime_error("Unsupported map type.");
   }
 }
 
-Properties::Properties(Properties&& o) noexcept
-  : keys(std::move(o.keys))
-  , values(std::move(o.values))
-{}
-
-Properties& Properties::operator=(Properties&& o) noexcept
+Properties::Properties(const AnyMap& p)
+  : props(p)
 {
-  keys = std::move(o.keys);
-  values = std::move(o.values);
-  return *this;
+  if (p.GetType() != AnyMap::UNORDERED_MAP_CASEINSENSITIVE_KEYS) {
+    props_check::ValidateAnyMap(p);
+
+    PopulateCaseInsensitiveLookupMap();
```

Comment:
Do the validation check as part of PopulateCaseInsensitiveLookupMap() (if faster...)